### PR TITLE
Implement basic book controller

### PR DESCRIPTION
**Pull request by RomanYuzviak**

### CODE REVIEW

File: src/main/java/com/example/onlinebookstore/OnlineBookStoreApplication.java [hunk 1]
```diff
@@ -1,13 +1,35 @@
...
+            Book newBook = new Book();
+            newBook.setAuthor("My Name");
+            newBook.setTitle("Some Book");
+            newBook.setCoverImage("Some cover image");
+            newBook.setDescription("Some description...");
+            newBook.setPrice(BigDecimal.valueOf(99));
+            newBook.setIsbn("09090994535");
+            //bookService.save(newBook);
+        };
+    }
```
Comment (Cayman4010) [inline src/main/java/com/example/onlinebookstore/OnlineBookStoreApplication.java]: As I remember we need to delete this bean

File: src/main/java/com/example/onlinebookstore/mapper/BookMapper.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package com.example.onlinebookstore.mapper;
+
+import com.example.onlinebookstore.config.MapperConfig;
+import com.example.onlinebookstore.dto.BookDto;
+import com.example.onlinebookstore.dto.CreateBookRequestDto;
+import com.example.onlinebookstore.model.Book;
+import org.mapstruct.Mapper;
+
+@Mapper(config = MapperConfig.class)
```
Comment (Cayman4010) [inline src/main/java/com/example/onlinebookstore/mapper/BookMapper.java]: you can add configuration just here, without additional config

File: src/main/java/com/example/onlinebookstore/repository/impl/BookRepositoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        }
+        return book;
+    }
+
+    @Override
+    public List<Book> findAll() {
+        try (Session session = sessionFactory.openSession()) {
+            return session.createQuery("FROM Book", Book.class).getResultList();
+        } catch (Exception ex) {
+            throw new RuntimeException("Can not get all books from database", ex);
```
Comment (Cayman4010) [inline src/main/java/com/example/onlinebookstore/repository/impl/BookRepositoryImpl.java]: You should use the custom EntityNotFoundException

File: src/main/java/com/example/onlinebookstore/repository/impl/BookRepositoryImpl.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        }
+    }
+
+    @Override
+    public Optional<Book> get(Long id) {
+        try (Session session = sessionFactory.openSession()) {
+            return Optional.ofNullable(
+                    session.get(Book.class, id));
+        } catch (Exception ex) {
+            throw new RuntimeException("Can not get book by id %d from database".formatted(id), ex);
```
Comment (Cayman4010) [inline src/main/java/com/example/onlinebookstore/repository/impl/BookRepositoryImpl.java]: here too

**Review by Cayman4010 — CHANGES_REQUESTED** [review]
Good job! But you can improve your code! :)

File: src/main/java/com/example/onlinebookstore/OnlineBookStoreApplication.java [hunk 5]
```diff
@@ -1,13 +1,16 @@
...
 
+import com.example.onlinebookstore.service.BookService;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 
 @SpringBootApplication
 public class OnlineBookStoreApplication {
+    @Autowired
+    private BookService bookService;
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/OnlineBookStoreApplication.java]: Do you need this service here?

File: src/main/java/com/example/onlinebookstore/mapper/BookMapper.java (hunk 2 above)
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/mapper/BookMapper.java]:
```suggestion
@Mapper(componentModel = "spring",
        injectionStrategy = InjectionStrategy.CONSTRUCTOR,
        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
        implementationPackage = "<PACKAGE_NAME>.impl")
```

File: src/main/java/com/example/onlinebookstore/mapper/BookMapper.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+public interface BookMapper {
+    BookDto toDto(Book book);
+
+    Book toModel(CreateBookRequestDto bookDto);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/mapper/BookMapper.java]: toBook will be better

File: src/main/java/com/example/onlinebookstore/model/Book.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Entity
+@Data
+public class Book {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/Book.java:15]: Add Table annotation

File: src/main/java/com/example/onlinebookstore/repository/BookRepository.java [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+import com.example.onlinebookstore.model.Book;
+import java.util.List;
+import java.util.Optional;
+
+public interface BookRepository {
+    Book save(Book book);
+
+    List<Book> findAll();
+
+    Optional<Book> get(Long id);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/repository/BookRepository.java]: bad naming
findById

File: src/main/java/com/example/onlinebookstore/repository/impl/BookRepositoryImpl.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+import org.hibernate.SessionFactory;
+import org.hibernate.Transaction;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public class BookRepositoryImpl implements BookRepository {
+    private final SessionFactory sessionFactory;
+
+    @Autowired
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/repository/impl/BookRepositoryImpl.java]: You don't need Autowired here as you have one constructor
Also, be consistent and you can use RequiredArgsConstructor annotation like in BookServiceImpl

File: src/main/java/com/example/onlinebookstore/repository/impl/BookRepositoryImpl.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        try {
+            session = sessionFactory.openSession();
+            transaction = session.beginTransaction();
+            session.persist(book);
+            transaction.commit();
+        } catch (Exception ex) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new EntityNotFoundException("Can not create the book: " + book, ex);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/repository/impl/BookRepositoryImpl.java]: It's better to create one more custom exception DataProcessingException and throw it here

File: src/main/java/com/example/onlinebookstore/repository/impl/BookRepositoryImpl.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    public List<Book> findAll() {
+        try (Session session = sessionFactory.openSession()) {
+            return session.createQuery("FROM Book", Book.class).getResultList();
+        } catch (Exception ex) {
+            throw new EntityNotFoundException("Can not get all books from database", ex);
+        }
+    }
+
+    @Override
+    public Optional<Book> get(Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/repository/impl/BookRepositoryImpl.java]: rename

File: src/main/java/com/example/onlinebookstore/service/BookService.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+import com.example.onlinebookstore.dto.BookDto;
+import com.example.onlinebookstore.dto.CreateBookRequestDto;
+import java.util.List;
+
+public interface BookService {
+    BookDto save(CreateBookRequestDto book);
+
+    List<BookDto> findAll();
+
+    BookDto get(Long id);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/BookService.java:12]:
```suggestion
    BookDto findById(Long id);
```

File: src/main/java/com/example/onlinebookstore/service/impl/BookServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    @Override
+    public List<BookDto> findAll() {
+        return bookRepository.findAll().stream()
+                .map(bookMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public BookDto get(Long id) {
+        return bookMapper.toDto(bookRepository.get(id).orElseThrow(() ->
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/impl/BookServiceImpl.java]: This code is really hard to read. Also bad idea to throw RuntimeException
```suggestion
        Book book = bookRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Can't find book with id " + id));
        return bookMapper.toDto(book);
```

File: src/main/java/com/example/onlinebookstore/dto/CreateBookRequestDto.java [hunk 14]
```diff
@@ -0,0 +1,14 @@
+package com.example.onlinebookstore.dto;
+
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Data
+public class CreateBookRequestDto {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/dto/CreateBookRequestDto.java]: use record from Java 17

File: src/main/java/com/example/onlinebookstore/dto/BookDto.java [hunk 15]
```diff
@@ -0,0 +1,15 @@
+package com.example.onlinebookstore.dto;
+
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Data
+public class BookDto {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/dto/BookDto.java]: use record from Java 17

File: src/main/java/com/example/onlinebookstore/controller/BookController.java [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping(value = "/api/books")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/controller/BookController.java]:
```suggestion
@RequestMapping(value = "/books")
```

File: src/main/java/com/example/onlinebookstore/config/MapperConfig.java [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+import org.mapstruct.InjectionStrategy;
+import org.mapstruct.NullValueCheckStrategy;
+
+@org.mapstruct.MapperConfig(
+        componentModel = "spring",
+        injectionStrategy = InjectionStrategy.CONSTRUCTOR,
+        nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS,
+        implementationPackage = "<PACKAGE_NAME>.impl"
+)
+public class MapperConfig {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/config/MapperConfig.java]: remove class as you will replace config to BookMapper

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
Everything seems to be fine, but you need to resolve the merge conflicts

**Pull request merged**